src/connection.rs

```diff
@@ -520,32 +520,64 @@ impl ActualConnection {
     }
 }
 
-pub fn connect(
+fn build_connection(
     connection_info: &ConnectionInfo,
     timeout: Option<Duration>,
 ) -> RedisResult<Connection> {
     let con = ActualConnection::new(&connection_info.addr, timeout)?;
```

Comment:
`ActualConnection::new` will create a new tcp/unix stream on every call so for authenticated connections the new code will open 2 or 3 connections (if it fails the first authentication attempt). Failing to authenticate shouldn't put the connection in a bad state so you should be able to use a single connection throughout.

Reply:
I tried using a single connection for 2nd auth, but it did not work. Are there any knobs to debug connection state?

Reply:
Seems like redis actually just shuts down the connection if a command isn't formatted correctly. I guess that makes sense in most cases since there is no way in general of knowing how to uphold the protocol...

Creating a new connection after authentication fails seems fine then, however you could pass the `Connection` as a paramter to `connect_auth` to ensure we only create a second connection if authentication fails.

Reply:
> Seems like redis actually just shuts down the connection if a command isn't formatted correctly. I guess that makes sense in most cases since there is no way in general of knowing how to uphold the protocol...

I don't have yet enough information about connection life cycle, but I think, that an invalid command should not drop connection.